File: Source/AlphaTab/Rendering/Effects/CrescendoEffectInfo.cs
using AlphaTab.Model;
using AlphaTab.Rendering.Glyphs;

namespace AlphaTab.Rendering.Effects
{
    class CrescendoEffectInfo : IEffectBarRendererInfo
    {
        public string EffectId { get { return "crescendo"; } }
        public bool HideOnMultiTrack { get { return false; } }
        public bool CanShareBand { get { return true; } }
        public EffectBarGlyphSizing SizingMode { get { return EffectBarGlyphSizing.GroupedOnBeatToEnd; } }

        public bool ShouldCreateGlyph(Settings settings, Beat beat)
        {
            return beat.Crescendo != CrescendoType.None;
        }

       
        public EffectGlyph CreateNewGlyph(BarRendererBase renderer, Beat beat)
        {
            return new CrescendoGlyph(0, 0, beat.Crescendo);
        }

        public bool CanExpand(Beat @from, Beat to)
        {
            return from.Crescendo == to.Crescendo;
        }
    }
}

File: src/rendering/effects/CrescendoEffectInfo.ts
import { Beat } from '@src/model/Beat';
import { CrescendoType } from '@src/model/CrescendoType';
import { BarRendererBase } from '@src/rendering/BarRendererBase';
import { EffectBarGlyphSizing } from '@src/rendering/EffectBarGlyphSizing';
import { CrescendoGlyph } from '@src/rendering/glyphs/CrescendoGlyph';
import { EffectGlyph } from '@src/rendering/glyphs/EffectGlyph';
import { EffectBarRendererInfo } from '@src/rendering/EffectBarRendererInfo';
import { Settings } from '@src/Settings';
import { NotationElement } from '@src/NotationSettings';

export class CrescendoEffectInfo extends EffectBarRendererInfo {
    public get notationElement(): NotationElement {
        return NotationElement.EffectCrescendo;
    }

    public get hideOnMultiTrack(): boolean {
        return false;
    }

    public get canShareBand(): boolean {
        return true;
    }

    public get sizingMode(): EffectBarGlyphSizing {
        return EffectBarGlyphSizing.GroupedOnBeatToEnd;
    }

    public shouldCreateGlyph(settings: Settings, beat: Beat): boolean {
        return beat.crescendo !== CrescendoType.None;
    }

    public createNewGlyph(renderer: BarRendererBase, beat: Beat): EffectGlyph {
        return new CrescendoGlyph(0, 0, beat.crescendo);
    }

    public canExpand(from: Beat, to: Beat): boolean {
        return from.crescendo === to.crescendo;
    }
}
